### ISSUE TITLE
IPv6 Compliance RFC4191: Processing Router Advertisement with Route Information Option (Host Only) [v6LC.2.2.23 Part J]

### ISSUE THREAD

**Issue by LiveFreeAndRoam** [systemd/systemd#31979]
### systemd version the issue has been seen with

255 + yuwata/systemd: commit 59b72d6f47160c74a1d7f44bc3e2373b4e14dd58 (HEAD -> network-next, origin/network-next)

### Used distribution

Ubuntu 23

### Linux kernel version used

6.5.0-17-generic

### CPU architectures issue was seen on

x86_64

### Component

systemd-networkd

#### Test v6LC.2.2.23: Stable addresses (Host Only) - Part J: Route Information Option with a Prefix Length of 0 and Lifetime set to 0.
**Purpose**: Verify that the HUT uses a Route Information Options to choose the next-hop.
**Reference**: [RFC 4191, Section 3.1](https://www.ietf.org/rfc/rfc4191.html#section-3.1)

The [definitive statement ](https://www.ietf.org/rfc/rfc4191.html#section-3.1:~:text=When%20a%20type,is%20already%20present.)in the RFC is:

>   When a type C host receives a Router Advertisement, it modifies its
   Routing Table as follows.  When processing a Router Advertisement, a
   type C host first updates a ::/0 route based on the Router Lifetime
   and Default Router Preference in the Router Advertisement message
   header.  Then as the host processes Route Information Options in the
   Router Advertisement message body, it updates its routing table for
   each such option.  The Router Preference and Lifetime values in a
   ::/0 Route Information Option override the preference and lifetime
   values in the Router Advertisement header.  Updating each route is
   done as follows.  A route is located in the Routing Table based on
   the prefix, prefix length, and next-hop router.  If the received
   route's lifetime is zero, the route is removed from the Routing Table
   if present.  If a route's lifetime is non-zero, the route is added to
   the Routing Table if not present and the route's lifetime and
   preference is updated if the route is already present.

#### Network Configuration
```
       TN2
        | 
    +---+---+
    |       |
   TR1     TR2
    |       |
    +---+---+
        |
       HUT
```

#### Procedure

| Step | Action | Expected Behavior of HUT<sup>(1)</sup> |
|--------|--------|--------|
| 1 | TR1 sends `RA rtr-pref=Medium Life=600 prefix=X`. | |
| 2 | TR2 sends `RA rtr-pref=Medium Life=600 prefix=X` `RIO: len=1 prefixlen=0 prf=High Life=600 prefix=::` |
| 3 | TR1 forwards an Echo Request from TN2 to the HUT. | The HUT should respond to the Echo Request with an Echo Reply with a next-hop of TR2. |
| 4 | TR2 sends `RA rtr-pref=Medium Life=600 prefix=X` `RIO: len=1 prefixlen=0 prf=High Life=0 prefix=::` |  |
| 5 | TR1 forwards an Echo Request from TN2 to the HUT. | The HUT should respond to the Echo Request with an Echo Reply with a next-hop of TR1. |

### Unexpected behaviour you saw

In Step 5, the HUT sent an echo reply via TR2 when it should have sent it to TR1, because in Step 4, TR2 advertised in RIO that its Lifetime=0.  Networkd evidently ignores the router preference Lifetime value of zero in TR2's RIO.

### Steps to reproduce the problem

[v6LC_2_2_23j.zip](https://github.com/systemd/systemd/files/14765990/v6LC_2_2_23j.zip)

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
enp0s9: NDISC: Received Router Advertisement: flags none preference medium lifetime 10min
enp0s9: NDISC: Invoking callback for 'router' event.
enp0s9: Requesting NDisc route (n/a): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
Setting '/proc/sys/net/ipv6/neigh/enp0s9/base_reachable_time_ms' to '30000'
No change in value '30000', suppressing write
Setting '/proc/sys/net/ipv6/neigh/enp0s9/retrans_time_ms' to '1000'
No change in value '1000', suppressing write
Setting '/proc/sys/net/ipv6/conf/enp0s9/hop_limit' to '64'
No change in value '64', suppressing write
enp0s9: Requesting NDisc route (n/a): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc address (n/a): 2001:2:0:1000:a00:27ff:fe1a:432/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: manage-temporary-address,no-prefixroute, scope: global
enp0s9: Setting SLAAC addresses and router.
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
enp0s9: Configuring NDisc route (requesting): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Configuring NDisc route (requesting): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Configuring NDisc address (requesting): 2001:2:0:1000:a00:27ff:fe1a:432/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: manage-temporary-address,no-prefixroute, scope: global
enp0s9: Received new NDisc route (configured): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: ndisc_check_ready(): SLAAC addresses and routes are not set.
enp0s9: Received new NDisc route (configured): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: ndisc_check_ready(): SLAAC addresses and routes are not set.
enp0s9: Received new NDisc address (configured): 2001:2:0:1000:a00:27ff:fe1a:432/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: tentative,manage-temporary-address,no-prefixroute, scope: global
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
enp0s9: ndisc_check_ready(): no SLAAC address is ready.
enp0s9: NDISC: Received Router Advertisement: flags none preference medium lifetime 10min
enp0s9: NDISC: Invoking callback for 'router' event.
enp0s9: Requesting NDisc route (n/a): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
Setting '/proc/sys/net/ipv6/neigh/enp0s9/base_reachable_time_ms' to '30000'
No change in value '30000', suppressing write
Setting '/proc/sys/net/ipv6/neigh/enp0s9/retrans_time_ms' to '1000'
No change in value '1000', suppressing write
Setting '/proc/sys/net/ipv6/conf/enp0s9/hop_limit' to '64'
No change in value '64', suppressing write
enp0s9: Found an existing route that conflicts with new route based on a received RA, removing.
enp0s9: Removing NDisc route (configured): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc route (configured,removing): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc address (configured): 2001:2:0:1000:a00:27ff:fe1a:432/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: manage-temporary-address,no-prefixroute, scope: global
enp0s9: Route prefix is ::/0, ignoring
enp0s9: Setting SLAAC addresses and router.
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
enp0s9: Configuring NDisc route (requesting): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Configuring NDisc route (requesting,configured,removing): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Configuring NDisc address (requesting,configured): 2001:2:0:1000:a00:27ff:fe1a:432/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: manage-temporary-address,no-prefixroute, scope: global
enp0s9: Forgetting removed NDisc route (configuring): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Received new NDisc route (configured): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Received remembered NDisc route (configured): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: ndisc_check_ready(): SLAAC addresses and routes are not set.
enp0s9: Received new NDisc route (configured): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: ndisc_check_ready(): SLAAC addresses and routes are not set.
enp0s9: ndisc_check_ready(): no SLAAC address is ready.
enp0s9: NDISC: No RA received before link confirmation timeout
enp0s9: NDISC: Invoking callback for 'timeout' event.
enp0s9: NDisc handler get timeout event
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
enp0s9: Received updated NDisc address (configured): 2001:2:0:1000:a00:27ff:fe1a:432/64 (valid for 18h 12min 13s, preferred for 18h 12min 13s), flags: manage-temporary-address,no-prefixroute, scope: global
enp0s9: SLAAC addresses and routes set.
enp0s9: link_check_ready(): IPv4LL:no DHCPv4:no DHCPv6:no DHCP-PD:no NDisc:yes
enp0s9: State changed: configuring -> configured
Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_34 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=23 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_34 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=24 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=25 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
enp0s9: Received new foreign route (configured): dst: 2001:2:0:1000:a00:27ff:fe1a:432/128, src: n/a, gw: n/a, prefsrc: n/a, table: local(255), priority: 0, proto: kernel, scope: global, type: local, flags: n/a

# 
# Step 4. RA with RIO: lieftime=0.  The misbehaviour occurs after receiving this RA.
# 
enp0s9: NDISC: Received Router Advertisement: flags none preference medium lifetime 10min
enp0s9: NDISC: Invoking callback for 'router' event.
enp0s9: Found an existing route that conflicts with new route based on a received RA, removing.
enp0s9: Removing NDisc route (configured): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a

# 
# The route is removed here, which seems correct behavior.
# 
enp0s9: Requesting NDisc route (configured,removing): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
Setting '/proc/sys/net/ipv6/neigh/enp0s9/base_reachable_time_ms' to '30000'
No change in value '30000', suppressing write
Setting '/proc/sys/net/ipv6/neigh/enp0s9/retrans_time_ms' to '1000'
No change in value '1000', suppressing write
Setting '/proc/sys/net/ipv6/conf/enp0s9/hop_limit' to '64'
No change in value '64', suppressing write
enp0s9: Found an existing route that conflicts with new route based on a received RA, removing.
enp0s9: Removing NDisc route (configured): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc route (configured,removing): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc address (configured): 2001:2:0:1000:a00:27ff:fe1a:432/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: manage-temporary-address,no-prefixroute, scope: global
enp0s9: Route prefix is ::/0, ignoring
enp0s9: Setting SLAAC addresses and router.
enp0s9: Configuring NDisc route (requesting,configured,removing): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Configuring NDisc route (requesting,configured,removing): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Configuring NDisc address (requesting,configured): 2001:2:0:1000:a00:27ff:fe1a:432/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: manage-temporary-address,no-prefixroute, scope: global
enp0s9: Forgetting removed NDisc route (configuring): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Forgetting removed NDisc route (configuring): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a

# 
# I suspect this illustrates the incorrect behavior.
# The route was configured, even though the RIO contains a route lifetime of zero.
# 
enp0s9: Received new NDisc route (configured): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a

# 
# I think the rest is ok.
# 
enp0s9: Received remembered NDisc route (configured): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: ndisc_check_ready(): SLAAC addresses and routes are not set.
enp0s9: Received new NDisc route (configured): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: ndisc_check_ready(): SLAAC addresses and routes are not set.
enp0s9: Received updated NDisc address (configured): 2001:2:0:1000:a00:27ff:fe1a:432/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: manage-temporary-address,no-prefixroute, scope: global
enp0s9: SLAAC addresses and routes set.

# 
# Cleanup phase
# 
enp0s9: NDISC: Received Router Advertisement: flags none preference medium lifetime 0
enp0s9: NDISC: Invoking callback for 'router' event.
enp0s9: Removing NDisc route (configured): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Found an existing route that conflicts with new route based on a received RA, removing.
enp0s9: Removing NDisc route (configured): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc route (configured,removing): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Setting SLAAC addresses and router.
enp0s9: Refuse to configure NDisc route with zero lifetime.
enp0s9: Forgetting removed NDisc route (n/a): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Forgetting removed NDisc route (n/a): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: NDISC: Received Neighbor Advertisement from fe80::200:10ff:fe10:1060: Router=yes, Solicited=no, Override=yes
enp0s9: NDISC: Invoking callback for 'neighbor' event.
enp0s9: NDISC: Received Router Advertisement: flags none preference medium lifetime 0
enp0s9: NDISC: Invoking callback for 'router' event.
enp0s9: Removing NDisc route (configured): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc route (n/a): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc route (n/a): dst: 2001:2:0:2000::/64, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Setting SLAAC addresses and router.
enp0s9: State changed: configured -> configuring
Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_34 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=26 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
enp0s9: Refuse to configure NDisc route with zero lifetime.
enp0s9: Refuse to configure NDisc route with zero lifetime.
enp0s9: NDISC: Received Neighbor Advertisement from fe80::200:10ff:fe10:1061: Router=yes, Solicited=no, Override=yes
enp0s9: NDISC: Invoking callback for 'neighbor' event.
enp0s9: Forgetting removed NDisc route (n/a): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1061, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
```

**Comment by yuwata** [comment]
@LiveFreeAndRoam Is this fixed by #31997?

**Comment by LiveFreeAndRoam** [comment]
> @LiveFreeAndRoam Is this fixed by #31997?

Yes, it is now  passing! Thank you @yuwata.

I tested with:

```
$ git log -n 1
commit 12a9241bf2e4cc60e3fa35de9d9b777a27a64a22 (HEAD -> network-ndisc-zero-lifetime, origin/network-ndisc-zero-lifetime)
Author: Yu Watanabe <watanabe.yu+github@gmail.com>
Date:   Thu Mar 28 13:25:50 2024 +0900

    network/radv: allow to announce prefix and route information with zero lifetime
    
    Setting zero lifetime for prefix or route information option is valid,
    and is used for announcing the prefix or route is invalidated.
    
    Fixes #26239.
```

**Comment by yuwata** [comment]
@LiveFreeAndRoam Thank you!
